structopt-derive/src/attrs.rs

```diff
@@ -536,16 +560,60 @@ impl Attrs {
         self.methods.iter().find(|m| m.name == name)
     }
 
-    pub fn methods(&self) -> TokenStream {
-        let methods = self
+    /// generate methods from attributes on top of struct or enum
+    pub fn top_level_methods(mut self) -> TokenStream {
+        match (self.no_version.clone(), self.version.clone()) {
+            (Some(no_version), Some(_)) => span_error!(
+                no_version.span(),
+                "`no_version` and `version = \"version\"` can't be used together"
+            ),
+
+            (None, Some((ident, version))) => {
+                self.push_str_method(ident.into(), version.into());
+            }
+
+            (None, None) => {
+                let name = Sp::call_site("version".to_string());
+                let arg = env::var("CARGO_PKG_VERSION").unwrap_or_else(|_|{
+                    call_site_error!("`CARGO_PKG_VERSION` environment variable is not defined, use `version = \"version\" to set it manually or `no_version` to not set it at all")
+                });
+                let arg = Sp::call_site(arg);
+                self.push_str_method(name, arg);
+            }
+
+            (Some(_), None) => {}
+        }
+
+        self.author
+            .clone()
+            .map(|(ident, arg)| self.push_str_method(ident.into(), arg.into()));
+
+        self.about
+            .clone()
+            .map(|(ident, arg)| self.push_str_method(ident.into(), arg.into()));
+
+        let methods: Vec<_> = self
+            .methods
+            .iter()
+            .map(|&Method { ref name, ref args }| quote!( .#name(#args) ))
+            .collect();
+
+        quote!( #(#methods)* )
+    }
+
+    /// generate methods on top of a field
+    pub fn field_methods(self) -> TokenStream {
+        let methods: Vec<_> = self
             .methods
             .iter()
-            .map(|&Method { ref name, ref args }| quote!( .#name(#args) ));
+            .map(|&Method { ref name, ref args }| quote!( .#name(#args) ))
+            .collect();
```

Comment:
I think you don't need to collect, quote take an IntoIterator.

Reply:
That is correct, the `.collect` is going to be removed